### PR TITLE
Accelerate AABB computation add debug output

### PR DESCRIPTION
**Pull request by zhx06**
## Summary
Vectorize no-collision loss & solver profiling

## Detailed description
- **Reason:** the no-overlap loss scored every pair in a Python loop (`O(n)²` calls/iter), dominating solve time.
- **Change:** score all pairs in one batched op (`NoCollisionLossStrategy.compute_loss_batched`); add opt-in `RelationSolverParams.profile` (solve time, pair count, ms/iter); add equivalence + gradient tests.
- **Impact:** loss/gradients unchanged; ~18×/40×/60× faster at 8/16/32 objects during benchmarking. No API change.

### CODE REVIEW

File: isaaclab_arena/relations/relation_loss_strategies.py [hunk 1]
```diff
@@ -437,6 +437,7 @@ def __init__(self, slope: float = 10.0, debug: bool = False):
         self.slope = slope
         self.debug = debug
 
+    # Production scores pairs through compute_loss_batched; this single-pair form is the test reference.
```
Comment (alexmillane) [inline isaaclab_arena/relations/relation_loss_strategies.py]: is there any way we could move this function into the test so it's not bloating the user facing files?

File: isaaclab_arena/relations/relation_loss_strategies.py [hunk 2]
```diff
@@ -501,6 +503,42 @@ def compute_loss(
...
+        """Overlap-volume no-overlap loss for boxes already reduced to world-space extents.
+
+        The subject box carries gradient; it is pushed off the obstacle box (expanded by clearance).
+
+        Args:
+            clearance_m: Minimum clearance between boxes in meters.
+            subject_min: World-space min extent of the subject box (..., 3).
+            subject_max: World-space max extent of the subject box (..., 3).
+            obstacle_min: World-space min extent of the obstacle box (..., 3).
+            obstacle_max: World-space max extent of the obstacle box (..., 3).
```
Comment (alexmillane) [inline isaaclab_arena/relations/relation_loss_strategies.py]: Do we know something more about the shape here?

Maybe `(N, 3)` where `N` is the number of boxes?

File: isaaclab_arena/relations/relation_loss_strategies.py [hunk 3]
```diff
@@ -501,6 +503,42 @@ def compute_loss(
...
+
+        Returns:
+            Loss tensor of shape matching the leading dimensions of the inputs.
```
Comment (alexmillane) [inline isaaclab_arena/relations/relation_loss_strategies.py]: add specification of the shape.

File: isaaclab_arena/relations/relation_loss_strategies.py [hunk 4]
```diff
@@ -501,6 +503,42 @@ def compute_loss(
...
+        overlap_x = interval_overlap_axis_loss(
+            subject_min[..., 0], subject_max[..., 0], obstacle_min[..., 0], obstacle_max[..., 0]
+        )
+        overlap_y = interval_overlap_axis_loss(
+            subject_min[..., 1], subject_max[..., 1], obstacle_min[..., 1], obstacle_max[..., 1]
+        )
+        overlap_z = interval_overlap_axis_loss(
+            subject_min[..., 2], subject_max[..., 2], obstacle_min[..., 2], obstacle_max[..., 2]
+        )
+        return self.slope * (overlap_x * overlap_y * overlap_z)
```
Comment (alexmillane) [inline isaaclab_arena/relations/relation_loss_strategies.py:505]: Nice! Clean!

File: isaaclab_arena/relations/relation_solver.py [hunk 5]
```diff
@@ -22,6 +23,15 @@
     from isaaclab_arena.assets.object_base import ObjectBase
 
 
+class _NoOverlapPair(NamedTuple):
+    """One directed overlap penalty: the subject box is pushed off the (detached) obstacle box."""
+
+    subject_min: torch.Tensor
+    subject_max: torch.Tensor
+    obstacle_min: torch.Tensor
+    obstacle_max: torch.Tensor
```
Comment (alexmillane) [inline isaaclab_arena/relations/relation_solver.py]: I think we're using `@dataclass(frozen=True)` as an alternative to the named tuple.

Suggestion to remove the leading underscore.

Suggestion to add a comment about the tensor shapes.

File: isaaclab_arena/relations/relation_solver.py [hunk 6]
```diff
@@ -169,59 +179,63 @@ def _compute_no_overlap_loss(
...
 
-        return total_loss
+        self._last_no_overlap_pair_count = len(pairs)
+        if not pairs:
+            return zero_loss
+
+        subject_min = torch.stack([p.subject_min for p in pairs], dim=0)
+        subject_max = torch.stack([p.subject_max for p in pairs], dim=0)
+        obstacle_min = torch.stack([p.obstacle_min for p in pairs], dim=0)
+        obstacle_max = torch.stack([p.obstacle_max for p in pairs], dim=0)
```
Comment (alexmillane) [inline isaaclab_arena/relations/relation_solver.py:235]: Maybe just add a comment here. Could change the variable name too to indicate that this is a batch.

File: isaaclab_arena/relations/relation_solver.py [hunk 7]
```diff
@@ -169,59 +179,63 @@ def _compute_no_overlap_loss(
...
-                if debug:
-                    print(f"  [NoOverlap] {child.name} vs {anchor.name}: loss={loss.mean().item():.6f}")
-                total_loss = total_loss + loss
+                anchor_min, anchor_max = extents[anchor]
+                # Anchor extents are already constant, so no detach is needed.
+                pairs.append(_NoOverlapPair(child_min, child_max, anchor_min, anchor_max))
+                pair_names.append((child.name, anchor.name))
 
-            # Against other non-anchors (unique pairs, both directions)
+        for i, child in enumerate(non_anchor_objects):
```
Comment (alexmillane) [inline isaaclab_arena/relations/relation_solver.py:215]: suggestion to comment what's going on  in this loop.

File: isaaclab_arena/relations/relation_solver.py [hunk 8]
```diff
@@ -169,59 +179,63 @@ def _compute_no_overlap_loss(
...
-                    child_bbox=child_bbox,
-                    parent_world_bbox=anchor_world_bbox,
-                )
-                if debug:
-                    print(f"  [NoOverlap] {child.name} vs {anchor.name}: loss={loss.mean().item():.6f}")
-                total_loss = total_loss + loss
+                anchor_min, anchor_max = extents[anchor]
+                # Anchor extents are already constant, so no detach is needed.
+                pairs.append(_NoOverlapPair(child_min, child_max, anchor_min, anchor_max))
+                pair_names.append((child.name, anchor.name))
```
Comment (alexmillane) [inline isaaclab_arena/relations/relation_solver.py]: Add comment about what's going on in this loop at the top.

**Review by alexmillane — APPROVED** [review]
Looks good. I just left some nits.

File: isaaclab_arena/relations/relation_loss_strategies.py (hunk 1 above)
Comment (zhx06) [reply isaaclab_arena/relations/relation_loss_strategies.py]: Done. It is moved to test.

File: isaaclab_arena/relations/relation_loss_strategies.py (hunk 3 above)
Comment (zhx06) [reply isaaclab_arena/relations/relation_loss_strategies.py]: Done, now in shape of `(num_pairs, batch_size)`.

File: isaaclab_arena/relations/relation_loss_strategies.py (hunk 4 above)
Comment (zhx06) [reply isaaclab_arena/relations/relation_loss_strategies.py:505]: Thanks!

File: isaaclab_arena/relations/relation_solver.py (hunk 5 above)
Comment (zhx06) [reply isaaclab_arena/relations/relation_solver.py]: Done. It is converted to `@dataclass(frozen=True)`, renamed to `NoOverlapPair`

File: isaaclab_arena/relations/relation_solver.py (hunk 6 above)
Comment (zhx06) [reply isaaclab_arena/relations/relation_solver.py:235]: Done. Comments added.

File: isaaclab_arena/relations/relation_solver.py (hunk 7 above)
Comment (zhx06) [reply isaaclab_arena/relations/relation_solver.py:215]: Comments added.

File: isaaclab_arena/relations/relation_loss_strategies.py (hunk 2 above)
Comment (zhx06) [reply isaaclab_arena/relations/relation_loss_strategies.py]: Done.  It is now specified  `(num_pairs, batch_size, 3)`

File: isaaclab_arena/relations/relation_solver.py (hunk 8 above)
Comment (zhx06) [reply isaaclab_arena/relations/relation_solver.py]: Comments added.
